release-notes/7.0/preview/api-diff/preview1/Microsoft.AspNetCore.App/7.0-preview1_Microsoft.Extensions.Configuration.md

```diff
@@ -0,0 +1,159 @@
++        public static T? Get<T>(this IConfiguration configuration);
+-        public static T Get<T>(this IConfiguration configuration, Action<BinderOptions> configureOptions);
++        public static T? Get<T>(this IConfiguration configuration, Action<BinderOptions>? configureOptions);
+-        public static object GetValue(this IConfiguration configuration, Type type, string key);
++        public static object? GetValue(this IConfiguration configuration, Type type, string key);
+-        public static object GetValue(this IConfiguration configuration, Type type, string key, object defaultValue);
++        public static object? GetValue(this IConfiguration configuration, Type type, string key, object defaultValue);
+-        public static T GetValue<T>(this IConfiguration configuration, string key);
++        public static T? GetValue<T>(this IConfiguration configuration, string key);
+-        public static T GetValue<T>(this IConfiguration configuration, string key, T defaultValue);
++        public static T? GetValue<T>(this IConfiguration configuration, string key, T defaultValue);
+     }
++    public readonly struct ConfigurationDebugViewContext
+     public static class ConfigurationExtensions {
+-        public static IConfigurationBuilder Add<TSource>(this IConfigurationBuilder builder, Action<TSource> configureSource) where TSource : IConfigurationSource, new();
++        public static IConfigurationBuilder Add<TSource>(this IConfigurationBuilder builder, Action<TSource>? configureSource) where TSource : IConfigurationSource, new();
+-        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration);
++        public static IEnumerable<KeyValuePair<string, string?>> AsEnumerable(this IConfiguration configuration);
+-        public static IEnumerable<KeyValuePair<string, string>> AsEnumerable(this IConfiguration configuration, bool makePathsRelative);
++        public static IEnumerable<KeyValuePair<string, string?>> AsEnumerable(this IConfiguration configuration, bool makePathsRelative);
+-        public static bool Exists(this IConfigurationSection section);
++        public static bool Exists(this IConfigurationSection? section);
+-        public static string GetConnectionString(this IConfiguration configuration, string name);
++        public static string? GetConnectionString(this IConfiguration configuration, string name);
+     }
+     public class ConfigurationKeyComparer : IComparer<string> {
+-        public int Compare(string x, string y);
++        public int Compare(string? x, string? y);
+     }
+     public sealed class ConfigurationManager : IConfiguration, IConfigurationBuilder, IConfigurationRoot, IDisposable {
+-        IDictionary<string, object> Microsoft.Extensions.Configuration.IConfigurationBuilder.Properties { get; }
++        IDictionary<string, object> IConfigurationBuilder.Properties { get; }
```

Comment:
What changed here?

Reply:
Maybe the ref file was manually modified, instead of autogenerated using the `dotnet build /t:GenerateReferenceSource` command?

I think `IConfigurationBuilder` should have showed up with the full name, not just the interface name.

@halter73 The `IConfigurationBuilder.Properties` property was introduced in this PR: https://github.com/dotnet/runtime/pull/55338/files

If what I mentioned is the root cause of this diff, you might want to correct that by autogenerating the ref file.

Reply:
For nullable annotations this ref file was manually modified (same for all other extensions ref files). 
When I run `build /t:GenerateReferenceSource` it doesn't generate nullable annotations, so I had to add them manually

Reply:
I see. I haven't yet grasped how to force `/t:GenerateReferenceSource` to substitute an existing ref cs file. Sometimes it does, sometimes it doesn't. You might have had the same problem as me.

It might be worth to still pursue my suggestion of re-generating the ref file from scratch, to make sure we're not accidentally missing anything.

Reply:
Since these APIs actually look like that in Preview1, I'll leave this as is.

I opened this issue to track the request: https://github.com/dotnet/runtime/issues/65913

https://github.com/dotnet/runtime/blob/6033ef2c56db79e68e13f2856f42d250399d927e/src/libraries/Microsoft.Extensions.Configuration/ref/Microsoft.Extensions.Configuration.cs#L32-L45

Reply:
> Since these APIs actually look like that in Preview1, I'll leave this as is.

I don't understand. Isn't there literally zero difference for a consumer of the API? If this is purely a formatting difference, it's not an API difference and shouldn't be included.

Reply:
Ok, removing it.

Reply:
Actually, not the whole type: I'll have to leave one nullability change I see there. The rest of the APIs with formatting difference, I'll remove.

Reply:
Thanks